man/man8/psf.8.in
'\" t
.\"     Title: psf
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 22 Dec 1999
.\"    Manual: 2.4.10
.\"    Source: 2.4.10
.\"  Language: English
.\"
.TH "PSF" "8" "22 Dec 1999" "2\&.4\&.10" "2.4.10"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
psf \- PostScript filter
.SH "SYNOPSIS"
.HP \w'\fBpsf\fR\fB\fR\ 'u
\fBpsf\fR\fB\fR [\-n\ \fIname\fR] [\-h\ \fIhost\fR] [\-w\ \fIwidth\fR] [\-l\ \fIlength\fR] [\-i\ \fIindent\fR] [\-c]
.SH "DESCRIPTION"
.PP
\fBpsf\fR
is an
\fBlpd\fR
filter for PostScript printing\&.
\fBpsf\fR
interprets the name it was called with to determine what filters to invoke\&. First, if the string ``pap\*(Aq\*(Aq appears anywhere in the name,
\fBpsf\fR
invokes
\fBpap\fR
to talk to a printer via AppleTalk\&. Next, if the string ``rev\*(Aq\*(Aq appears,
\fBpsf\fR
invokes
\fBpsorder\fR
to reverse the pages of the job\&. Finally, if
\fBpsf\fR
was called with a filter\*(Aqs name as the leading string, it invokes that filter\&. If there is no filter to run,
\fBpsf\fR
examines the magic number of the input, and if the input is not PostScript, converts it to PostScript\&.
.SH "KLUDGE"
.PP
In the default configuration,
\fBpsf\fR
supports two kludges\&. The first causes
\fBpsf\fR
to check its name for the letter `m\*(Aq\&. If this letter is found and accounting is turned on,
\fBpsf\fR
calls
\fBpap\fR
twice, once to get an initial page count and to print the job, and another time to get a final page count\&. This is a work\-around for bugs in a variety of PAP implementions that cause printers to never properly close the PAP output file\&. A notable example is any printer by Hewlett\-Packard\&.
.PP
The second kludge causes
\fBpsf\fR
to examine its name for the letter `w\*(Aq\&. If this letter is found and accounting is turned on,
\fBpsf\fR
calls
\fBpap\fR
with the
\fB\-w\fR
flag\&. This flag causes
\fBpap\fR
to wait until the printer\*(Aqs status contains the string `idle\*(Aq\&. Once this string is found, the job is printed as normal\&. This kludge is a work\-around for printers, notably Hewlett\-Packard\*(Aqs LaserJet IV, which will report a page count while a previous jobs is still printing\&.
.SH "EXAMPLE"
.PP
The sample
\fBprintcap\fR
entry below invokes
\fBpsf\fR
to print text files, PostScript files,
\fBtroff\fR\*(Aqs C/A/T output, and
\fBTeX\fR\*(Aqs DVI output, to an AppleTalk connected LaserWriter Plus\&. Since the LaserWriter Plus stacks pages in descending order, we reverse the pages and print the burst page last\&.
.sp
.if n \{\
.RS 4
.\}
.nf
laser|lp|LaserWriter Plus on AppleTalk:\e
    :sd=/var/spool/lpd/laser:\e
    :lp=/var/spool/lpd/laser/null:\e
    :lf=/var/log/lpd\-errs:pw#80:hl:\e
    :of=ofpap:\e
    :if=ifpaprev:\e
    :tf=tfpaprev:\e
    :df=dfpaprev:
.fi
.if n \{\
.RE
.\}
.PP
Note that if the host in question spools to more than one AppleTalk printer,
/dev/null
should not be used for the
\fBlp\fR
capability\&. Instead, a null device should be created with
\fBmknod\fR
for each printer, as has been done above\&.
.PP
Finally, there is a file in the spool directory,
/var/spool/lpd/laser, called
\fB\&.paprc\fR, which
\fBpap\fR
reads for the AppleTalk name of the printer\&.
.SH "SEE ALSO"
.PP
\fBpsorder\fR(1),
\fBprintcap\fR(5),
\fBlpd\fR(1),
\fBmknod\fR(1),
\fBpap\fR(1)\&.


NEWS
Changes in 4.0.0
================
* NEW: Reintroduce AppleTalk / DDP support, GitHub #220
       Controlled with the new build system option `-Dwith-appletalk'.
       Revived daemons: atalkd, papd, timelord, a2boot
       Revived config files: atalkd.conf, papd.conf
       Revived utilities: aecho, getzones, nbplkup, nbprgstr, nbpunrgstr, pap,
       papstatus
* NEW: Bundle macipgw, the MacIP Gateway daemon by Stefan Bethke, GitHub #1204
* UPD: uams: All encrypted UAMs depend on Libgcrypt now, GitHub #1488, #1506
       This means we remove the bundled wolfSSL library.
       A big thanks to the wolfSSL team for all their support!
* FIX: uams: Remove unhelpful Libgcrypt version check, GitHub #1550
* UPD: Remove the obsoleted PGP UAM, GitHub #1507
* NEW: Bundle, configure and install the Webmin module, GitHub #518
       Controlled with the new build system option `-Dwith-webmin'.
* UPD: Migrate afpstats from dbus-glib to GDBus, GitHub #666
       Special thanks to Simon McVittie for his help!
* UPD: Remove canned troff man pages from distribution, GitHub #460
       The build system now generates them on the fly.
       Introduces a build time dependency on DocBook XSL and xsltproc.
* UPD: Remove generated Unicode conversion tables, GitHub #1220
       Introduces a build time dependency on the UnicodeData.txt database.
* UPD: Detect host OS home dir and configure afp.conf on the fly, GitHub #1274
* UPD: meson: Autodetect init style for host OS, #1124
* UPD: meson: Allow building with multiple init styles, GitHub #1291
* NEW: meson: Introduce `-Dwith-readmes' option for installing additional docs.
       GitHub #1310
* UPD: Remove the Autotools build system. Meson is now the only choice.
       GitHub #1213

Changes in 3.2.10
=================
* BREAKING: Install netatalk-dbus.conf into datadir by default, GitHub #1533
       Previously: sysconfdir. This can be overridden by the build system.
* FIX: uams: Correct shadow password length check for ClearTxt, GitHub #1528
* FIX: cnid_dbd: Set explicit max length of db_params to prevent potential
       buffer overflow, GitHub #694
* FIX: meson: Debugging was enabled by default causing tickles
       to not be sent out, GitHub #1514
* FIX: meson: Format afpd help text output to match autotools, GitHub #1499
* FIX: meson: Throw missing cracklib dictionary warning, GitHub #1495
* FIX: meson: Use a valid code sample for the TCP Wrappers check, GitHub #1491

Changes in 3.2.9
================
* UPD: Use the recommended command to import Solaris init manifest,
       GitHub #1451
* FIX: uams: Make sure the DHX2 client nonce is aligned appropriately,
       GitHub #1456
* FIX: uams: Fix DHCAST128 key alignment problem, GitHub #1464
* FIX: wolfssl: OpenSSL coexistence tweaks, GitHub #1469
* FIX: docs: Remove straggler path substitution in afp.conf, GitHub #1480

Changes in 3.2.8
================
* UPD: Bump bundled WolfSSL library to stable version 5.7.2, GitHub #1433
       Resolves CVE-2024-1544, CVE-2024-5288, CVE-2024-5991, CVE-2024-5814
* UPD: Revert local modifications to the bundled WolfSSL library, GitHub #1432
* FIX: Enable building against a shared WolfSSL 5.7.2 library, GitHub #1421
* FIX: meson: Do not define rpath with a linker argument, GitHub #1443

Changes in 3.2.7
================
* NEW: meson: Ability to control the run-time linker path config file,
       GitHub #1396
       New boolean Meson option: `-Dwith-ldsoconf'
       When set to false, do not create /etc/ld.so.conf.d/libatalk.conf
* BREAKING: meson: Enable rpath by default, while disabling ldsoconf
       by default, GitHub #1417
* FIX: meson: Allow ldconfig to run unprivileged during setup, GitHub #1407
* FIX: docker: Add entry script step to clean up any residual lock file,
       GitHub #1412
* NEW: docker: Ship a docker-compose.yml sample file, GitHub #1414
* NEW: docker: Check for AFP_USER and AFP_PASS when launching container,
       GitHub #1415

Changes in 3.2.6
================
* BREAKING: meson: Refresh the dynamic linker cache when installing on Linux,
       GitHub #1386
       This fixes the issue of the libatalk.so shared library not being found
       when configuring with a non-standard library path, e.g. /usr/local/lib .
       New Meson option `-Dwith-install-hooks' controls this behavior,
       allowing you to disable the install hook in non-privileged environments.
       On Linux systems with glibc, we now install the following config file:
       /etc/ld.so.conf.d/libatalk.conf
* BREAKING: meson: Introduce option to control which manual l10n to build,
       GitHub #1390
       New Meson option `-Dwith-manual-l10n' default to empty, can be set to
       `ja' to build the Japanese localization of the html manual.
       This changes the default behavior of the build system
       to not build the Japanese html manual by default.
* BREAKING: meson: Install htmldocs into htmldocs subdir, GitHub #1391
       Previously, the html manual files were installed into the root
       of the netatalk doc directory. Now they are put under netatalk/htmldocs .
* BREAKING: meson: Use modern linker flag for rpath, remove dtags override,
       GitHub #1384
       When configuring with `-Dwith-rpath=true' the linker flags
       `-Wl,-rpath,' will be prepended instead of the old `-R' flag.
       On Linux platforms, we no longer prepend `-Wl,--enable-new-dtags',
       either.

Changes in 3.2.5
================
* BREAKING: meson: Allow choosing shared or static libraries to build,
       GitHub #1321
       In practice, only shared libraries are built by default now.
       Use the `default_library' option to control what is built.
* FIX: meson: Control the MySQL CNID backend, and support MariaDB, GitHub #1341
       Introduces a new boolean `with-cnid-mysql-backend' option.
* FIX: meson: Implement with-init-dir option, GitHub #1346
* FIX: autotools/meson: Install FreeBSD init script into correct location,
       GitHub #1345

etc/psf/psf.c
/*
 * Copyright (c) 1990,1995 Regents of The University of Michigan.
 * All Rights Reserved. See COPYRIGHT.
 *
 * PostScript Filter, psf.
 *
 * Handles both PostScript files and text files.  Files with the
 * '%!' PostScript header are sent directly to the printer,
 * unmodified.  Text files are first converted to PostScript,
 * then sent.  Printers may be directly attached or on an AppleTalk
 * network.  Other media are possible.  Currently, psf invokes
 * pap to send files to AppleTalk-ed printers.  Replace the pap*
 * variables to use another program for communication.  psf only
 * converts plain-text.  If called as "tf" or "df", psf will invoke
 * a troff or dvi to PostScript converter.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#define FUCKED

#include <unistd.h>
#include <sys/time.h>

/* POSIX.1 sys/wait.h check */
#include <sys/types.h>
#ifdef HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif /* HAVE_SYS_WAIT_H */
#ifndef WEXITSTATUS
#define WEXITSTATUS(stat_val) ((unsigned)(stat_val) >> 8)
#endif /* ! WEXITSTATUS */
#ifndef WIFEXITED
#define WIFEXITED(stat_val) (((stat_val) & 255) == 0)
#endif /* ! WIFEXITED */

#include <sys/file.h>
#include <syslog.h>
#include <atalk/paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <errno.h>

/* Forward Declarations */
int pexecv(char *path, char *argv[]);
int copyio();
int textps();

static char		psapath[] = _PATH_PSA;
static char		*psaargv[] = { "psa", NULL, NULL, NULL, NULL };

/*
 * If we're not doing accounting, we just call pap as below.
 * If we are doing accounting, we call pap twice.  The first time,
 * we call it with "-E" in arg 2, pagecount.ps in arg 3, and "-" in
 * arg 4.  The second time, we call it with "-c" in arg 2, pagecount.ps
 * in arg 3, and 0 in arg 4.
 */
static char		pappath[] = _PATH_PAP;
static char		*papargv[] = { "pap", "-sstatus", NULL, NULL, NULL, NULL, NULL, NULL };

static char		revpath[] = _PATH_PSORDER;
static char		*revargv[] = { "psorder", "-d", NULL };

static char		*filtargv[] = { NULL, NULL, NULL };

static char		inbuf[ 1024 * 8 ];
static int		inlen;

static int		literal;
static int		width = 80, length = 66, indent = 0;
static char		*prog, *name, *host;

static struct papersize {
    int		width;
    int		length;
   float	win;
    float	lin;
} papersizes[] = {
    { 80, 66, 8.5, 11.0 },			/* US Letter */
    { 80, 70, 8.27, 11.69 },			/* A4 */
};

int main( int ac, char **av)
{
    int			c, rc, children = 0;
#ifdef FUCKED
    int			psafileno = 0, multiconn = 0, waitidle = 0, waitidle2 = 0;
#endif /* FUCKED */
    int			status;
    extern char		*optarg;
    extern int		optind, opterr;

    opterr = 0;
    if (( prog = rindex( av[ 0 ], '/' )) == NULL ) {
	prog = av[ 0 ];
    } else {
	prog++;
    }
    openlog( prog, LOG_PID, LOG_LPR );

    while (( c = getopt( ac, av, "P:C:D:F:L:J:x:y:n:h:w:l:i:c" )) != EOF ) {
	switch ( c ) {
	case 'n' :
	    name = optarg;
	    break;

	case 'h' :
	    host = optarg;
	    break;

	case 'w' :
	    width = atoi( optarg );
#ifdef ZEROWIDTH
	    /*
	     * Some version of lpd pass 0 for the page width.
	     */
	    if ( width == 0 ) {
		width = 80;
	    }
#endif /* ZEROWIDTH */
	    break;

	case 'l' :
	    length = atoi( optarg );
	    break;

	case 'i' :
	    indent = atoi( optarg );
	    break;

	case 'c' :	/* Print control chars */
	    literal++;
	    break;

	case 'F' :
	case 'L' :
	case 'J' :
	case 'P' :
	case 'x' :
	case 'y' :
	    break;
	
#ifdef notdef
	default :
	    syslog( LOG_ERR, "bad option: %c", c );
	    exit( 2 );
#endif /* notdef */
	}
    }
    if ( ac - optind > 1 ) {
	syslog( LOG_ERR, "Too many arguments" );
	exit( 2 );
    }
#ifdef FUCKED
    if ( index( prog, 'w' )) {
	waitidle++;
    }
    if ( index( prog, 'W' )) {
	waitidle2++;
    }
    if ( index( prog, 'm' )) {
	multiconn++;
    }
#endif /* FUCKED */

    syslog( LOG_INFO, "starting for %s", name ? name : "?" );

restart:
    if (( inlen = read( 0, inbuf, sizeof( inbuf ))) < 0 ) {
	syslog( LOG_ERR, "read: %s", strerror(errno) );
	exit( 1 );
    }
    if ( inlen == 0 ) {	/* nothing to be done */
	syslog( LOG_INFO, "done" );
	exit( 0 );
    }

    /*
     * If we've been given an accounting file, start the accounting
     * process.
     */
    if ( optind < ac ) {
	/* build arguments */
	psaargv[ 1 ] = av[ optind ];
	psaargv[ 2 ] = name;
	psaargv[ 3 ] = host;
	if (( c = pexecv( psapath, psaargv )) < 0 ) {
	    syslog( LOG_ERR, "%s: %s", psapath, strerror(errno) );
	    exit( 2 );
	}
	children++;
	syslog( LOG_INFO, "accounting with psa[%d]", c );
    }

    /*
     * Check prog's name to decide what programs to execute.
     */
    if ( strstr( prog, "pap" ) != NULL ) {
	if ( optind < ac ) {	/* accounting */
#ifdef FUCKED
	    if ( multiconn ) {
		psafileno = getdtablesize();
		psafileno--;
		dup2( 1, psafileno );

		if ( waitidle2 ) {
		    papargv[ 2 ] = "-W";
		    papargv[ 3 ] = "-c";
		    papargv[ 4 ] = "-E";
		    papargv[ 5 ] = _PATH_PAGECOUNT;
		    papargv[ 6 ] = "-";
		    papargv[ 7 ] = NULL;
		} else if ( waitidle ) {
		    papargv[ 2 ] = "-w";
		    papargv[ 3 ] = "-c";
		    papargv[ 4 ] = "-E";
		    papargv[ 5 ] = _PATH_PAGECOUNT;
		    papargv[ 6 ] = "-";
		    papargv[ 7 ] = NULL;
		} else {
		    papargv[ 2 ] = "-c";
		    papargv[ 3 ] = "-E";
		    papargv[ 4 ] = _PATH_PAGECOUNT;
		    papargv[ 5 ] = "-";
		    papargv[ 6 ] = NULL;
		}
	    } else {
		/*
		 * This is how it should be done.
		 */
		papargv[ 2 ] = "-c";
		papargv[ 3 ] = _PATH_PAGECOUNT;
		papargv[ 4 ] = "-";
		papargv[ 5 ] = _PATH_PAGECOUNT;
		papargv[ 6 ] = NULL;
	    }
#endif /* FUCKED */
	} else {
	    papargv[ 2 ] = "-c";
	    papargv[ 3 ] = "-E";
	    papargv[ 4 ] = NULL;
	}

	if (( c = pexecv( pappath, papargv )) < 0 ) {
	    syslog( LOG_ERR, "%s: %s", pappath, strerror(errno) );
	    exit( 2 );
	}
	children++;
	syslog( LOG_INFO, "sending to pap[%d]", c );
    }

    /*
     * Might be a good idea to have both a "forw" and a "rev", so that
     * reversed documents can be reordered for the printing device.
     */
    if ( strstr( prog, "rev" ) != NULL ) {
	if (( c = pexecv( revpath, revargv )) < 0 ) {
	    syslog( LOG_ERR, "%s: %s", revpath, strerror(errno) );
	    exit( 2 );
	}
	syslog( LOG_INFO, "sending to rev[%d]", c );
	children++;
    }

    /*
     * Invoke an external (script) filter to produce PostScript from
     * non-text input.
     */
    if ( *prog != 'i' && *prog != 'o' && *( prog + 1 ) == 'f' ) {
	filtargv[ 0 ] = filtargv[ 1 ] = prog;
	if (( c = pexecv( _PATH_PSFILTER, filtargv )) < 0 ) {
	    syslog( LOG_ERR, "%s: %s", _PATH_PSFILTER, strerror(errno) );
	    exit( 2 );
	}
	syslog( LOG_INFO, "external filter[%d]", c );
	children++;
	rc = copyio();		/* external filter */
    } else {
	if ( inlen >= 2 && inbuf[ 0 ] == '%' && inbuf[ 1 ] == '!' ) {
	    syslog( LOG_INFO, "PostScript" );
	    rc = copyio();	/* PostScript */
	} else if ( inlen >= 2 && inbuf[ 0 ] == '\033' && inbuf[ 1 ] == '%' ) {
	    syslog( LOG_INFO, "PostScript w/PJL" );
	    rc = copyio();      /* PostScript */
	} else {
	    syslog( LOG_INFO, "straight text" );
	    rc = textps();	/* straight text */
	}
    }

#ifdef FUCKED
    if ( strstr( prog, "pap" ) != NULL && optind < ac && multiconn ) {
	dup2( psafileno, 1 );
	close( psafileno );
	papargv[ 2 ] = "-c";
	if ( waitidle2 ) {
	    papargv[ 3 ] = "-W";
	    papargv[ 4 ] = _PATH_PAGECOUNT;
	    papargv[ 5 ] = NULL;
	} else if ( waitidle ) {
	    papargv[ 3 ] = "-w";
	    papargv[ 4 ] = _PATH_PAGECOUNT;
	    papargv[ 5 ] = NULL;
	} else {
	    papargv[ 3 ] = _PATH_PAGECOUNT;
	    papargv[ 4 ] = NULL;
	}

	if (( c = pexecv( pappath, papargv )) < 0 ) {
	    syslog( LOG_ERR, "%s: %s", pappath, strerror(errno) );
	    exit( 2 );
	}
	children++;
	syslog( LOG_INFO, "pagecount with pap[%d]", c );
    }
#endif /* FUCKED */

    if ( children ) {
	close( 1 );
    }
    while ( children ) {
	if (( c = wait3( &status, 0, NULL )) < 0 ) {
	    syslog( LOG_ERR, "wait3: %s", strerror(errno) );
	    exit( 1 );
	}
	if ( WIFEXITED( status )) {
#ifndef WEXITSTATUS
#define WEXITSTATUS(x)	((x).w_status)
#endif /* WEXITSTATUS */
	    if ( WEXITSTATUS( status ) != 0 ) {
		syslog( LOG_ERR, "%d died with %d", c, WEXITSTATUS( status ));
		exit( WEXITSTATUS( status ));
	    } else {
		syslog( LOG_INFO, "%d done", c );
		children--;
	    }
	} else {
	    syslog( LOG_ERR, "%d died badly", c );
	    exit( 1 );
	}
    }

    if ( rc == 3 ) {
	syslog( LOG_INFO, "pausing" );
	kill( getpid(), SIGSTOP );
	syslog( LOG_INFO, "restarting" );
	goto restart;
    }

    syslog( LOG_INFO, "done" );
    exit( rc );
}

int copyio(void)
{
    /* implement the FSM needed to do the suspend. Note that
     * the last characters will be \031\001 so don't worry
     * Fun things: 1. \031\001 should not be written to output device
     *  2. The \031 can be last char of one read, \001 first of next
     *      - we need to write \031 if not followed by \001
     */
    struct timeval	tv;
    fd_set		fdset;
    int			ctl = 0;

notdone:
    do {
	/*
	 * First, \031 and \001 *must* be the last things in the buffer
	 * (\001 can be the first thing in the next buffer).  There's no
	 * need to scan any of the intervening bytes.  Second, if there's
	 * more input, the escape sequence was bogus, and we should keep
	 * reading.
	 */
	if ( inlen == 1 ) {
	    if ( ctl == 1 ) {
		if ( inbuf[ 0 ] == '\001' ) {
		    ctl = 2;
		    break;
		}
		if ( write( 1, "\031", 1 ) != 1 ) {
		    syslog( LOG_ERR, "write: %s", strerror(errno) );
		    return( 1 );
		}
		ctl = 0;
	    }

	    if ( inbuf[ 0 ] == '\031' ) {
		ctl = 1;
	    }

	} else {
	    if ( ctl == 1 ) {
		if ( write( 1, "\031", 1 ) != 1 ) {
		    syslog( LOG_ERR, "write: %s", strerror(errno) );
		    return( 1 );
		}
	    }
	    ctl = 0;
	    if ( inbuf[ inlen - 2 ] == '\031' &&
		    inbuf[ inlen - 1 ] == '\001' ) {
		ctl = 2;
	    } else if ( inbuf[ inlen - 1 ] == '\031' ) {
		ctl = 1;
	    }
	}

	inlen -= ctl;
	if (( inlen > 0 ) && ( write( 1, inbuf, inlen ) != inlen )) {
	    syslog( LOG_ERR, "write: %s", strerror(errno) );
	    return( 1 );
	}
	if ( ctl == 2 ) {
	    break;
	}
    } while (( inlen = read( 0, inbuf, sizeof( inbuf ))) > 0 );

    if ( ctl == 2 ) {
	tv.tv_sec = 2;
	tv.tv_usec = 0;
	FD_ZERO( &fdset );
	FD_SET( 0, &fdset );
	if ( select( 1, &fdset, NULL, NULL, &tv ) != 0 ) {
	    if (( inlen = read( 0, inbuf, sizeof( inbuf ))) > 0 ) {
		goto notdone;
	    }
	}
    }

    if ( inlen < 0 ) {
	syslog( LOG_ERR, "read: %s", strerror(errno) );
	return( 1 );
    }

    if ( ctl == 1 ) {
	if ( write( 1, "\031", 1 ) != 1 ) {
	    syslog( LOG_ERR, "write: %s", strerror(errno) );
	    return( 1 );
	}
    } else if ( ctl == 2 ) {
	return( 3 );
    }
    return( 0 );
}

static char		*font = "Courier";
static int		point = 11;

static char		pspro[] = "\
/GSV save def						% global VM\n\
/SP {\n\
	/SV save def					% save vmstate\n\
	dup /H exch def					% save font height\n\
	exch findfont exch scalefont setfont		% select font\n\
	( ) stringwidth pop /W exch def			% save font width\n\
	0.5 sub 72 mul /CY exch def			% save start Y\n\
	pop 0.5 add 72 mul /CX exch def			% save start X\n\
	CX CY moveto					% make current point\n\
} bind def\n\
/S /show load def\n\
/NL { CX CY H sub dup /CY exch def moveto } bind def\n\
/CR { CX CY moveto } bind def\n\
/B { W neg 0 rmoveto}bind def\n\
/T { W mul 0 rmoveto}bind def\n\
/EP { SV restore showpage } bind def\n\
%%EndProlog\n";

int textps(void)
{
    struct papersize	papersize;
    int			state = 0, line = 0, col = 0, npages = 0, rc;
    unsigned int	i;
    char		*p, *end;

#define elements(x)	(sizeof(x)/sizeof((x)[0]))
    for ( i = 0; i < elements( papersizes ); i++ ) {
	if ( width == papersizes[ 0 ].width &&
		length == papersizes[ 0 ].length ) {
	    papersize = papersizes[ i ];
	    break;
	}
    }
    if ( i >= elements( papersizes )) {
	papersize = papersizes[ 0 ];		/* default */
    }

#define ST_AVAIL		(1<<0)
#define ST_CONTROL		(1<<1)
#define ST_PAGE			(1<<2)
    /*
     * convert text lines to postscript.
     * A grungy little state machine. If I was more creative, I could
     * probably think of a better way of doing this...
     */
    do {
	p = inbuf;
	end = inbuf + inlen;
	while ( p < end ) {
	    if (( state & ST_PAGE ) == 0 && *p != '\031' && *p != '\001' ) {
		if ( npages == 0 ) {
		    printf( "%%!PS-Adobe-2.0\n%%%%Pages: (atend)\n" );
		    printf( "%%%%DocumentFonts: %s\n", font );
		    fflush( stdout );

		    /* output postscript prologue: */
		    if ( write( 1, pspro, sizeof( pspro ) - 1 ) !=
			    sizeof( pspro ) - 1 ) {
			syslog( LOG_ERR, "write prologue: %s", strerror(errno) );
			return( 1 );
		    }
		    if ( name && host ) {
			printf( "statusdict /jobname (%s@%s) put\n", name,
				host );
		    }
		}

		printf( "%%%%Page: ? %d\n", ++npages );
		printf( "%d %f %f /%s %d SP\n", indent,
			papersize.win, papersize.lin, font, point );
		state |= ST_PAGE;
	    }
	    if ( state & ST_CONTROL && *p != '\001' ) {
		/* It is a very bad thing to toss a job because it contains
		 * unprintable characters.  Instead, we will convert them to
		 * question marks.  This is adapted from a solution described
		 * by Werner Eugster <eugster@giub.unibe.ch> on his ApplePrint
		 * webpage (http://www.giub.unibe.ch/~eugster/appleprint.html).
		 *
		 * Note that this is rather ugly code.  The same change is
		 * applied identically at two different locations in this file.
		 * It would be better someday to combine the two.
                */
		if ( !literal ) {
			fprintf( stderr,
				"unprintable character (0x%x) converted to ?!\n",
				(unsigned char)*p );
			putchar( '?' ); /* Replace unprintable char with a question mark. */
		} else {
			printf( "\\%o", (unsigned char)031 );
		}
		state &= ~ST_CONTROL;
		col++;
	    }

	    switch ( *p ) {
	    case '\n' :		/* end of line */
		if ( state & ST_AVAIL ) {
		    printf( ")S\n" );
		    state &= ~ST_AVAIL;
		}
		printf( "NL\n" );
		line++;
		col = 0;
		if ( line >= length ) {
		    printf( "EP\n" );
		    state &= ~ST_PAGE;
		    line = 0;
		}
		break;

	    case '\r' :		/* carriage return (for overtyping) */
		if ( state & ST_AVAIL ) {
		    printf( ")S CR\n" );
		    state &= ~ST_AVAIL;
		}
		col = 0;
		break;

	    case '\f' :		/* form feed */
		if ( state & ST_AVAIL ) {
		    printf( ")S\n" );
		    state &= ~ST_AVAIL;
		}
		printf( "EP\n" );
		state &= ~ST_PAGE;
		line = 0;
		col = 0;
		break;

	    case '\b' :		/* backspace */
		/* show line, back up one character */
		if ( state & ST_AVAIL ) {
		    printf( ")S\n" );
		    state &= ~ST_AVAIL;
		}
		printf( "B\n" );
		col--;
		break;

	    case '\t' :		/* tab */
		if ( state & ST_AVAIL ) {
		    printf( ")S\n" );
		    state &= ~ST_AVAIL;
		}
		printf( "%d T\n", 8 - ( col % 8 ));
		col += 8 - ( col % 8 );
		break;

	    /*
	     * beginning of lpr control sequence
	     */
	    case '\031' :
		state |= ST_CONTROL;
		break;

	    case '\001' :	/* lpr control sequence */
		if ( state & ST_CONTROL ) {
		    rc = 3;
		    goto out;
		}
		/* FALLTHROUGH */

	    case '\\' :
	    case ')' :
	    case '(' :
		if (( state & ST_AVAIL ) == 0 ) {
		    printf( "(" );
		    state |= ST_AVAIL;
		}
		putchar( '\\' );
		/* FALLTHROUGH */

	    default :
		if (( state & ST_AVAIL ) == 0 ) {
		    printf( "(" );
		    state |= ST_AVAIL;
		}
		if ( !isascii( *p ) || !isprint( *p )) {
		    if ( !literal ) {
                        fprintf( stderr,
			    "unprintable character (0x%x) converted to ?!\n",
			    (unsigned char)*p );
			putchar( '?' ); /* Replace unprintable char with a question mark. */
		    } else {
		        printf( "\\%o", (unsigned char)*p );
		    }
		} else {
		    putchar( *p );
		}
		col++;
		break;
	    }
	p++;
	}
    } while (( inlen = read( 0, inbuf, sizeof( inbuf ))) > 0 );
    if ( inlen < 0 ) {
	syslog( LOG_ERR, "read: %s", strerror(errno) );
	return( 1 );
    }
    rc = 0;

out:
    if ( state & ST_AVAIL ) {
	printf( ")S\n" );
	state &= ~ST_AVAIL;
    }

    if ( state & ST_PAGE ) {
	printf( "EP\n" );
	state &= ~ST_PAGE;
    }

    if ( npages > 0 ) {
	printf( "%%%%Trailer\nGSV restore\n%%%%Pages: %d\n%%%%EOF\n", npages );
	fflush( stdout );
    }

    return( rc );
}

/*
 * Interface to pipe and exec, for starting children in pipelines.
 *
 * Manipulates file descriptors 0, 1, and 2, such that the new child
 * is reading from the parent's output.
 */
int pexecv( char *path, char *argv[])
{
    int		fd[ 2 ], c;

    if ( pipe( fd ) < 0 ) {
	return( -1 );
    }

    switch ( c = fork()) {
    case -1 :
	return( -1 );
	/* NOTREACHED */

    case 0 :
	if ( close( fd[ 1 ] ) < 0 ) {
	    return( -1 );
	}
	if ( dup2( fd[ 0 ], 0 ) < 0 ) {
	    return( -1 );
	}
	if ( close( fd[ 0 ] ) < 0 ) {
	    return( -1 );
	}
	execv( path, argv );
	return( -1 );
	/* NOTREACHED */

    default :
	if ( close( fd[ 0 ] ) < 0 ) {
	    return( -1 );
	}
	if ( dup2( fd[ 1 ], 1 ) < 0 ) {
	    return( -1 );
	}
	if ( close( fd[ 1 ] ) < 0 ) {
	    return( -1 );
	}
	return( c );
    }
}
